<a href="https://colab.research.google.com/github/OsirisValencia/InteligenciaArtificialUdeA/blob/main/02_preprocesado.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = '.'
!kaggle
!chmod 600 ./kaggle.json

usage: kaggle [-h] [-v] [-W]
              {competitions,c,datasets,d,kernels,k,models,m,files,f,config}
              ...
kaggle: error: the following arguments are required: command


In [ ]:
!pip install nbformat --quiet
import nbformat

ruta = "/content/drive/MyDrive/Colab Notebooks/02_preprocesado.ipynb"  # actualiza esta ruta
nb = nbformat.read(ruta, as_version=4)

# 1) Borra widgets de cada celda (por si quedaba alguno)
for cell in nb.cells:
    if "widgets" in cell.metadata:
        del cell.metadata["widgets"]

# 2) Borra widgets del metadata de nivel raíz
if "widgets" in nb.metadata:
    del nb.metadata["widgets"]

# Guarda por encima del mismo archivo
nbformat.write(nb, ruta)
print("✓ He eliminado todos los metadata.widgets (celdas + raíz).")


✓ He eliminado todos los metadata.widgets (celdas + raíz).


In [ ]:
!kaggle competitions download -c udea-ai-4-eng-20251-pruebas-saber-pro-colombia
!unzip udea-ai-4-eng-20251-pruebas-saber-pro-colombia.zip

udea-ai-4-eng-20251-pruebas-saber-pro-colombia.zip: Skipping, found more recently modified local copy (use --force to force download)
Archive:  udea-ai-4-eng-20251-pruebas-saber-pro-colombia.zip
replace submission_example.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: submission_example.csv  
replace test.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: test.csv                
replace train.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: train.csv               


In [ ]:
import pandas as pd
import numpy as np


df = pd.read_csv('train.csv')

# Guardamos el target antes de droppear
y = df['RENDIMIENTO_GLOBAL'].copy()

# Drop de ID y columnas duplicadas / irrelevantes
df = df.drop(['ID','FAMI_TIENEINTERNET.1','RENDIMIENTO_GLOBAL'], axis=1)


In [ ]:
# Primeras columnas
df.head(20).T

0   \
PERIODO                                                            20212   
ESTU_PRGM_ACADEMICO                                           ENFERMERIA   
ESTU_PRGM_DEPARTAMENTO                                            BOGOTÁ   
ESTU_VALORMATRICULAUNIVERSIDAD  Entre 5.5 millones y menos de 7 millones   
ESTU_HORASSEMANATRABAJA                                Menos de 10 horas   
FAMI_ESTRATOVIVIENDA                                           Estrato 3   
FAMI_TIENEINTERNET                                                    Si   
FAMI_EDUCACIONPADRE                     Técnica o tecnológica incompleta   
FAMI_TIENELAVADORA                                                    Si   
FAMI_TIENEAUTOMOVIL                                                   Si   
ESTU_PRIVADO_LIBERTAD                                                  N   
ESTU_PAGOMATRICULAPROPIO                                              No   
FAMI_TIENECOMPUTADOR                                                  Si   
FAMI_EDUCACIONMADRE                                            Postgrado   
coef_1                                                             0.322   
coef_2                                                             0.208   
coef_3                                                              0.31   
coef_4                                                             0.267   

                                                                      1   \
PERIODO                                                            20212   
ESTU_PRGM_ACADEMICO                                              DERECHO   
ESTU_PRGM_DEPARTAMENTO                                         ATLANTICO   
ESTU_VALORMATRICULAUNIVERSIDAD  Entre 2.5 millones y menos de 4 millones   
ESTU_HORASSEMANATRABAJA                                                0   
FAMI_ESTRATOVIVIENDA                                           Estrato 3   
FAMI_TIENEINTERNET                                                    No   
FAMI_EDUCACIONPADRE                       Técnica o tecnológica completa   
FAMI_TIENELAVADORA                                                    Si   
FAMI_TIENEAUTOMOVIL                                                   No   
ESTU_PRIVADO_LIBERTAD                                                  N   
ESTU_PAGOMATRICULAPROPIO                                              No   
FAMI_TIENECOMPUTADOR                                                  Si   
FAMI_EDUCACIONMADRE                     Técnica o tecnológica incompleta   
coef_1                                                             0.311   
coef_2                                                             0.215   
coef_3                                                             0.292   
coef_4                                                             0.264   

                                                                      2   \
PERIODO                                                            20203   
ESTU_PRGM_ACADEMICO                                MERCADEO Y PUBLICIDAD   
ESTU_PRGM_DEPARTAMENTO                                            BOGOTÁ   
ESTU_VALORMATRICULAUNIVERSIDAD  Entre 2.5 millones y menos de 4 millones   
ESTU_HORASSEMANATRABAJA                                  Más de 30 horas   
FAMI_ESTRATOVIVIENDA                                           Estrato 3   
FAMI_TIENEINTERNET                                                    Si   
FAMI_EDUCACIONPADRE                   Secundaria (Bachillerato) completa   
FAMI_TIENELAVADORA                                                    Si   
FAMI_TIENEAUTOMOVIL                                                   No   
ESTU_PRIVADO_LIBERTAD                                                  N   
ESTU_PAGOMATRICULAPROPIO                                              No   
FAMI_TIENECOMPUTADOR                                                  No   
FAMI_EDUCACIONMADRE                   Secundaria (Bachillerato) completa   
coef_1                                  

In [ ]:

pd.DataFrame({'Columna': df.columns, 'Valores Distintos': [df[col].nunique() for col in df.columns]})

Columna  Valores Distintos
0                          PERIODO                  9
1              ESTU_PRGM_ACADEMICO                948
2           ESTU_PRGM_DEPARTAMENTO                 31
3   ESTU_VALORMATRICULAUNIVERSIDAD                  8
4          ESTU_HORASSEMANATRABAJA                  5
5             FAMI_ESTRATOVIVIENDA                  7
6               FAMI_TIENEINTERNET                  2
7              FAMI_EDUCACIONPADRE                 12
8               FAMI_TIENELAVADORA                  2
9              FAMI_TIENEAUTOMOVIL                  2
10           ESTU_PRIVADO_LIBERTAD                  2
11        ESTU_PAGOMATRICULAPROPIO                  2
12            FAMI_TIENECOMPUTADOR                  2
13             FAMI_EDUCACIONMADRE                 12
14                          coef_1                375
15                          coef_2                342
16                          coef_3                196
17                          coef_4                276

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

def visualizar_columnas_con_pocos_valores(dataframe, max_valores=13):
    # Identificar columnas con menos de max_valores valores únicos
    columnas_filtradas = [col for col in dataframe.columns if dataframe[col].nunique() < max_valores]

    if not columnas_filtradas:
        print(f"No hay columnas con menos de {max_valores} valores únicos.")
        return

    print(f"Se encontraron {len(columnas_filtradas)} columnas con menos de {max_valores} valores únicos.\n")

    # Para cada columna, crear una visualización
    for i, columna in enumerate(columnas_filtradas):
        # Obtener conteo de valores
        conteo = dataframe[columna].value_counts(dropna=False)

        df_conteo = pd.DataFrame({
            'Valor': conteo.index.astype(str),
            'Cantidad': conteo.values,
            'Porcentaje': (conteo.values / len(dataframe) * 100).round(2)

        })

        # Ordenar por cantidad (mayor a menor)
        df_conteo = df_conteo.sort_values('Cantidad', ascending=False).reset_index(drop=True)

        # Imprimir tabla de distribución
        print(f"\n--- Columna: {columna} ({conteo.count()} valores únicos) ---")
        print(df_conteo.to_string(index=False))


# Ejemplo de uso
visualizar_columnas_con_pocos_valores(df)

Se encontraron 12 columnas con menos de 13 valores únicos.


--- Columna: PERIODO (9 valores únicos) ---
Valor  Cantidad  Porcentaje
20195    180873       26.12
20203    171838       24.81
20212    171412       24.75
20183    164818       23.80
20194      1472        0.21
20213      1178        0.17
20202       490        0.07
20184       254        0.04
20196       165        0.02

--- Columna: ESTU_VALORMATRICULAUNIVERSIDAD (9 valores únicos) ---
                                   Valor  Cantidad  Porcentaje
  Entre 1 millón y menos de 2.5 millones    204048       29.47
Entre 2.5 millones y menos de 4 millones    127430       18.40
                        Menos de 500 mil     80263       11.59
       Entre 500 mil y menos de 1 millón     78704       11.37
Entre 4 millones y menos de 5.5 millones     69736       10.07
                       Más de 7 millones     68014        9.82
Entre 5.5 millones y menos de 7 millones     38490        5.56
                       No pagó matrícula   

# Proceso de limpieza correspondiente.

Primero debemos definir qué tipo de dato tienen las columnas, si es un tipo de dato categorico, discreto, continuo...
Esto nos ayudará a definir una estrategía para saber cómo rellenar los valores faltantes.
Inicialmente el dataset todos los tipos de datos son objects, vamos a "Normalizarlos" siguiendo los diferentes tipos de datos. por ejemplo algunos solo tienen 2 posibles valores si y no, los trataremos como binarios. hay algunos valores discretos como lo puede ser ESTRATO_VIVIENDA y otros continuos como lo puede ser valormatriculauniversidad que a su vez puede ser categorico.
<br/>
Anteriormente identificamos las siguientes columnas con valores faltantes:




| Tipo                     | Columnas                                                                                                                                                         | Nº categorías/distintos |
| ------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------- |
| **Continuas**            | coef\_1, coef\_2, coef\_3, coef\_4                                                                                                                               | 375, 342, 196, 276      |
| **Ordinales**            | ESTU\_VALORMATRICULAUNIVERSIDAD (8+1 nan), ESTU\_HORASSEMANATRABAJA (5+1 nan), FAMI\_ESTRATOVIVIENDA (7+1 nan), PERIODO (9 valores, pero solo **primeros 4 dígitos** relevantes: año + posible semestre)                                                   | 8, 5, 7, 4                 |
| **Binarias (`Si`/`No`)** | FAMI\_TIENEINTERNET, FAMI\_TIENELAVADORA, FAMI\_TIENEAUTOMOVIL, ESTU\_PRIVADO\_LIBERTAD, ESTU\_PAGOMATRICULAPROPIO, FAMI\_TIENECOMPUTADOR, FAMI\_TIENEINTERNET.1 | todas tienen 2+1 nan    |
| **Multiclase altas**     | ESTU\_PRGM\_ACADEMICO (948), ESTU\_PRGM\_DEPARTAMENTO (31), FAMI\_EDUCACIONPADRE (12+1 nan), FAMI\_EDUCACIONMADRE (12+1 nan)                                     | 948, 31, 12, 12         |                                                                       | 9                       |
| **Target**               | RENDIMIENTO\_GLOBAL (4 clases balanceadas)                                                                                                                       | 4                       |


**Tratamiento de PERIODO** Cada valor es un entero de 5 dígitos como 20195, pero solo nos interesa los 4 primeros indican año Propuesta:

Convertir a string y quedarnos con los 4 primeros caracteres:

*Estrategias propuestas para la inputacion de datos según el tipo de datos que haya en la columna*

| Tipo de variable           | Estrategia recomendada                                                                                                                                  |
| -------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Continuas**              | – Imputación por **mediana** (robusta frente a outliers).<br>– Alternativa: KNNImputer si hay correlaciones.                                            |
| **Ordinales**              | – Imputar con **valor modal** (la categoría más frecuente).<br>– O crear categoría “Desconocido” para conservar missing.                                |
| **Binarias (`Si`/`No`)**   | – Imputar con la **frecuencia** de `Si`/`No` (modo).<br>– O codificar `nan` como categoría propia (`-1` o `Unknown`).                                   |
| **Multiclase altas**       | – Dado el alto cardinal, imputar con categoría “Desconocido” para no distorsionar.<br>– O usar técnicas basadas en similitud (e.g. MostFrequent / KNN). |         |


| Tipo                                              | Columnas & estrategia                                                              |
| ------------------------------------------------- | ---------------------------------------------------------------------------------- |
| **Continuas**                                     | – **StandardScaler** (media=0, var=1) para coef\_\*.                               |
| **Ordinales**                                     | – **OrdinalEncoder** respetando el orden natural:<br>  - ESTU\_HORASSEMANATRABAJA: |
| `["0","Menos de 10","11-20","21-30","Más de 30"]` |                                                                                    |


In [ ]:
df['PERIODO_STR'] = df['PERIODO'].astype(str).str.zfill(5)
df['SEMESTRE'] = df['PERIODO_STR'].str[:4].astype(int)
df = df.drop(['PERIODO','PERIODO_STR'], axis=1)

In [ ]:
from sklearn.compose        import ColumnTransformer
from sklearn.impute         import SimpleImputer, KNNImputer
from sklearn.preprocessing  import StandardScaler, OneHotEncoder, OrdinalEncoder


# Definición de columnas
numeric_cols    = ["coef_1","coef_2","coef_3","coef_4"]
ordinal_cols    = ["SEMESTRE",
                   "ESTU_HORASSEMANATRABAJA",
                   "FAMI_ESTRATOVIVIENDA",
                   "ESTU_VALORMATRICULAUNIVERSIDAD"]
binary_cols     = ["FAMI_TIENEINTERNET", "FAMI_TIENELAVADORA",
                   "FAMI_TIENEAUTOMOVIL","ESTU_PRIVADO_LIBERTAD",
                   "ESTU_PAGOMATRICULAPROPIO","FAMI_TIENECOMPUTADOR"]
highcard_cols   = ["FAMI_EDUCACIONPADRE","FAMI_EDUCACIONMADRE"]  # excluyendo el programa y la región

# Imputación de valores faltantes
#    - Numéricas: mediana
num_imp = SimpleImputer(strategy='median')
df[numeric_cols] = num_imp.fit_transform(df[numeric_cols])

#  3)- Ordinales: moda (most_frequent), categorizando nan como 'Desconocido'
ord_imp = SimpleImputer(strategy='most_frequent', fill_value='Desconocido')
df[ordinal_cols] = ord_imp.fit_transform(df[ordinal_cols])

#  4) Binarias: moda (most_frequent), categorizando nan como 'Desconocido'
bin_imp = SimpleImputer(strategy='most_frequent', fill_value='Desconocido')
df[binary_cols] = bin_imp.fit_transform(df[binary_cols])

# 5) Columnas de alta cardinalidad
high_imp = SimpleImputer(strategy='constant', fill_value='Desconocido')

# Ajuste y transformación
df[highcard_cols] = high_imp.fit_transform(df[highcard_cols])

# 5) Mostrar resultados
print("Primeras filas tras imputación:")
print(df.head())

print("\nConteo de valores nulos por columna:")
print(df.isnull().sum())

print("\nDistribución de categorías (incluyendo 'Desconocido'):")
for col in highcard_cols:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False).head(10))

Primeras filas tras imputación:
          ESTU_PRGM_ACADEMICO ESTU_PRGM_DEPARTAMENTO  \
0                  ENFERMERIA                 BOGOTÁ   
1                     DERECHO              ATLANTICO   
2       MERCADEO Y PUBLICIDAD                 BOGOTÁ   
3  ADMINISTRACION DE EMPRESAS              SANTANDER   
4                  PSICOLOGIA              ANTIOQUIA   

             ESTU_VALORMATRICULAUNIVERSIDAD ESTU_HORASSEMANATRABAJA  \
0  Entre 5.5 millones y menos de 7 millones       Menos de 10 horas   
1  Entre 2.5 millones y menos de 4 millones                       0   
2  Entre 2.5 millones y menos de 4 millones         Más de 30 horas   
3  Entre 4 millones y menos de 5.5 millones                       0   
4  Entre 2.5 millones y menos de 4 millones     Entre 21 y 30 horas   

  FAMI_ESTRATOVIVIENDA FAMI_TIENEINTERNET                 FAMI_EDUCACIONPADRE  \
0            Estrato 3                 Si    Técnica o tecnológica incompleta   
1            Estrato 3                 No 

In [ ]:
from pandas.api.types import CategoricalDtype

# Definir los órdenes para variables ordinales
horas_order = ["0", "Menos de 10 horas", "Entre 11 y 20 horas",
               "Entre 21 y 30 horas", "Más de 30 horas", "Desconocido"]
estrato_order = ["Estrato 1", "Estrato 2", "Estrato 3",
                 "Estrato 4", "Estrato 5", "Estrato 6", "Sin Estrato", "Desconocido"]
matri_order = ["Menos de 500 mil", "Entre 500 mil y menos de 1 millón",
               "Entre 1 millón y menos de 2.5 millones",
               "Entre 2.5 millones y menos de 4 millones",
               "Entre 4 millones y menos de 5.5 millones",
               "Entre 5.5 millones y menos de 7 millones",
               "Más de 7 millones", "No pagó matrícula", "Desconocido"]

# 2) Crear CategoricalDtype para cada ordinal
ord_horas = CategoricalDtype(categories=horas_order, ordered=True)
ord_estrato = CategoricalDtype(categories=estrato_order, ordered=True)
ord_matri = CategoricalDtype(categories=matri_order, ordered=True)

# 3) Asignar dtypes
df = df.astype({
    'FAMI_EDUCACIONPADRE': 'category',
    'FAMI_EDUCACIONMADRE': 'category',
    'FAMI_TIENEINTERNET': 'category',
    'FAMI_TIENELAVADORA': 'category',
    'FAMI_TIENEAUTOMOVIL': 'category',
    'ESTU_PRIVADO_LIBERTAD': 'category',
    'ESTU_PAGOMATRICULAPROPIO': 'category',
    'FAMI_TIENECOMPUTADOR': 'category',
    'SEMESTRE': 'int16',
    'coef_1': 'float64',
    'coef_2': 'float64',
    'coef_3': 'float64',
    'coef_4': 'float64',
})

# 4) Aplicar categorías ordenadas
df['ESTU_HORASSEMANATRABAJA'] = df['ESTU_HORASSEMANATRABAJA'].astype(ord_horas)
df['FAMI_ESTRATOVIVIENDA'] = df['FAMI_ESTRATOVIVIENDA'].astype(ord_estrato)
df['ESTU_VALORMATRICULAUNIVERSIDAD'] = df['ESTU_VALORMATRICULAUNIVERSIDAD'].astype(ord_matri)

# 5) Verificación
print(df.dtypes)


ESTU_PRGM_ACADEMICO                 object
ESTU_PRGM_DEPARTAMENTO              object
ESTU_VALORMATRICULAUNIVERSIDAD    category
ESTU_HORASSEMANATRABAJA           category
FAMI_ESTRATOVIVIENDA              category
FAMI_TIENEINTERNET                category
FAMI_EDUCACIONPADRE               category
FAMI_TIENELAVADORA                category
FAMI_TIENEAUTOMOVIL               category
ESTU_PRIVADO_LIBERTAD             category
ESTU_PAGOMATRICULAPROPIO          category
FAMI_TIENECOMPUTADOR              category
FAMI_EDUCACIONMADRE               category
coef_1                             float64
coef_2                             float64
coef_3                             float64
coef_4                             float64
SEMESTRE                             int16
dtype: object


Hasta el momento no hemos tratado las columnas ESTU_PRGM_ACADEMICO                 
ESTU_PRGM_DEPARTAMENTO
por la cantidad de valores diferentes que hay, veamos qué podemos hacer.

In [ ]:
!pip install unidecode

In [ ]:
import unicodedata, re
def normalize_text(s):
    s = str(s).lower()
    s = unicodedata.normalize('NFKD', s).encode('ASCII', 'ignore').decode('ASCII')
    s = re.sub(r'[^a-z0-9 ]+', ' ', s)
    return re.sub(r'\s+', ' ', s).strip()

df['PRGM_NORM'] = df['ESTU_PRGM_ACADEMICO'].map(normalize_text)


In [ ]:
!pip install rapidfuzz

In [ ]:
from rapidfuzz import fuzz
from scipy.cluster.hierarchy import linkage, fcluster

# 1) lista única
progs = df['PRGM_NORM'].unique().tolist()

# 2) matriz de similitud (sparse si es necesario)
N = len(progs)
D = np.zeros((N, N))
for i in range(N):
    for j in range(i+1, N):
        score = fuzz.token_sort_ratio(progs[i], progs[j]) / 100
        D[i,j] = D[j,i] = 1 - score

# 3) linkage y corte
Z = linkage(D, method='average')
labels = fcluster(Z, t=0.3, criterion='distance')  # t ajusta sensibilidad

# 4) mapping de programa → clúster
cluster_map = dict(zip(progs, labels))
df['PRGM_CLUSTER'] = df['PRGM_NORM'].map(cluster_map)


<ipython-input-48-10a83ec6d5da>:17: ClusterWarning: The symmetric non-negative hollow observation matrix looks suspiciously like an uncondensed distance matrix
  Z = linkage(D, method='average')


In [ ]:
from sentence_transformers import SentenceTransformer
from sklearn.cluster import KMeans

progs = df['PRGM_NORM'].unique().tolist()

model = SentenceTransformer('all-MiniLM-L6-v2')
embs  = model.encode(progs, show_progress_bar=True)
kmeans = KMeans(n_clusters=80).fit(embs)
cluster_map_sem = dict(zip(progs, kmeans.labels_))
df['PRGM_CLUSTER_SEM'] = df['PRGM_NORM'].map(cluster_map_sem)


Batches:   0%|          | 0/24 [00:00<?, ?it/s]

In [ ]:
from sklearn.preprocessing import OneHotEncoder
mapping = {
    'ingenieria': 'Ingeniería y Tecnologías',
    'sistemas': 'Ingeniería y Tecnologías',
    'medicina': 'Ciencias de la Salud y Veterinaria',
    'enfermeria': 'Ciencias de la Salud y Veterinaria',
    'veterinaria': 'Ciencias de la Salud y Veterinaria',
    'administracion': 'Negocios, Economía y Derecho',
    'finanzas': 'Negocios, Economía y Derecho',
    'derecho': 'Negocios, Economía y Derecho',
    'psicologia': 'Ciencias Sociales y Educación',
    'social': 'Ciencias Sociales y Educación',
    'pedagogia': 'Ciencias Sociales y Educación',
    'comunicacion': 'Artes y Comunicación',
    'musica': 'Artes y Comunicación',
    'diseno': 'Artes y Comunicación',
    'biologia': 'Ciencias Naturales y Ambientales',
    'quimica': 'Ciencias Naturales y Ambientales',
    'ambiental': 'Ciencias Naturales y Ambientales',
    'licenciatura': 'Educación y Pedagogía'
}
def map_facultad(s):
    for kw, fac in mapping.items():
        if kw in s:
            return fac
    return 'Otros/Interdisciplinarios'
df['FACULTAD'] = df['PRGM_NORM'].map(map_facultad)

# Normalizamos acentos en departamento
df['DEP_NORM'] = (
    df['ESTU_PRGM_DEPARTAMENTO']
      .astype(str)
      .str.normalize('NFKD')
      .str.encode('ASCII', 'ignore').str.decode('ASCII')
      .str.lower()
      .str.strip()
)
# --- 4) Mapeo a REGIONES (6 categorías) ---
region_map = {
  **dict.fromkeys([d.lower() for d in [
      'Antioquia','Cundinamarca','Boyaca','Tolima',
      'Quindio','Risaralda','Santander','Norte de Santander',
      'Huila','Caldas']], 'Andina'),
  **dict.fromkeys([d.lower() for d in [
      'Atlantico','Bolivar','Magdalena','Cesar',
      'Cordoba','Sucre','La Guajira']], 'Caribe'),
  **dict.fromkeys([d.lower() for d in ['Choco','Valle del Cauca','Cauca','Narino']], 'Pacífica'),
  **dict.fromkeys([d.lower() for d in ['Meta','Arauca','Casanare','Vichada']], 'Orinoquia'),
  **dict.fromkeys([d.lower() for d in [
      'Amazonas','Caqueta','Guainia','Guaviare',
      'Putumayo','Vaupes']], 'Amazonía'),
  'San Andres y Providencia': 'Insular'
}

df['REGION'] = df['DEP_NORM'].map(region_map).fillna('otros')

# --- 5) One-Hot Encoding para FACULTAD y REGION ---
ohe = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
encoded = ohe.fit_transform(df[['FACULTAD', 'REGION']])
encoded_cols = ohe.get_feature_names_out(['FACULTAD', 'REGION'])

df = pd.concat([
    df.reset_index(drop=True),
    pd.DataFrame(encoded, columns=encoded_cols)
], axis=1).drop(['ESTU_PRGM_ACADEMICO','ESTU_PRGM_DEPARTAMENTO',
                'PRGM_NORM','DEP_NORM','FACULTAD','REGION'], axis=1)

# Verifica las columnas restantes
print("Columnas restantes:")
print(df.columns.tolist())

# --- 6) Resultado ---
print("Nuevas columnas One-Hot añadidas:")
print(encoded_cols)


Columnas restantes:
['ESTU_VALORMATRICULAUNIVERSIDAD', 'ESTU_HORASSEMANATRABAJA', 'FAMI_ESTRATOVIVIENDA', 'FAMI_TIENEINTERNET', 'FAMI_EDUCACIONPADRE', 'FAMI_TIENELAVADORA', 'FAMI_TIENEAUTOMOVIL', 'ESTU_PRIVADO_LIBERTAD', 'ESTU_PAGOMATRICULAPROPIO', 'FAMI_TIENECOMPUTADOR', 'FAMI_EDUCACIONMADRE', 'coef_1', 'coef_2', 'coef_3', 'coef_4', 'SEMESTRE', 'PRGM_CLUSTER', 'PRGM_CLUSTER_SEM', 'FACULTAD_Artes y Comunicación', 'FACULTAD_Ciencias Naturales y Ambientales', 'FACULTAD_Ciencias Sociales y Educación', 'FACULTAD_Ciencias de la Salud y Veterinaria', 'FACULTAD_Educación y Pedagogía', 'FACULTAD_Ingeniería y Tecnologías', 'FACULTAD_Negocios, Economía y Derecho', 'FACULTAD_Otros/Interdisciplinarios', 'REGION_Amazonía', 'REGION_Andina', 'REGION_Caribe', 'REGION_Orinoquia', 'REGION_Pacífica', 'REGION_otros']
Nuevas columnas One-Hot añadidas:
['FACULTAD_Artes y Comunicación'
 'FACULTAD_Ciencias Naturales y Ambientales'
 'FACULTAD_Ciencias Sociales y Educación'
 'FACULTAD_Ciencias de la Salud y Vet

In [ ]:
df.shape

(692500, 32)

Ahora vamos a normalizar las columnas faltantes para que luego sea posible en la siguiente entrega evaluarlas con algún modelo de clasificación.

In [ ]:
from sklearn.preprocessing import OrdinalEncoder

matri_order = [
  'Menos de 500 mil',            # 0
  'Entre 500 mil y menos de 1 millón',
  'Entre 1 millón y menos de 2.5 millones',
  'Entre 2.5 millones y menos de 4 millones',
  'Entre 4 millones y menos de 5.5 millones',
  'Entre 5.5 millones y menos de 7 millones',
  'Más de 7 millones',           # 6
  'No pagó matrícula',           # 7
  'Desconocido'                  # 8 (si aplica)
]

horas_order = [
  '0', 'Menos de 10 horas',
  'Entre 11 y 20 horas',
  'Entre 21 y 30 horas',
  'Más de 30 horas',
  'Desconocido'
]

estrato_order = [
  'Estrato 1','Estrato 2','Estrato 3',
  'Estrato 4','Estrato 5','Estrato 6',
  'Sin Estrato','Desconocido'
]

ord_enc_matri = OrdinalEncoder(
    categories=[matri_order], dtype=int,
    handle_unknown='use_encoded_value', unknown_value=-1
)
df['VALOR_MATRICULA_ENC'] = ord_enc_matri.fit_transform(
    df[['ESTU_VALORMATRICULAUNIVERSIDAD']]
)

ord_enc_horas = OrdinalEncoder(
    categories=[horas_order], dtype=int,
    handle_unknown='use_encoded_value', unknown_value=-1
)
df['HORAS_TRABAJO_ENC'] = ord_enc_horas.fit_transform(
    df[['ESTU_HORASSEMANATRABAJA']]
)

ord_enc_estrato = OrdinalEncoder(
    categories=[estrato_order], dtype=int,
    handle_unknown='use_encoded_value', unknown_value=-1
)
df['ESTRATO_ENC'] = ord_enc_estrato.fit_transform(
    df[['FAMI_ESTRATOVIVIENDA']]
)

# El Ordinal Encoding preserva el orden y consume solo una columna
# El Ordinal Encoding es óptimo para ordinales porque reduce dimensionalidad y mantiene el orden

In [ ]:
# Variables binarias


bin_map = {'No': 0, 'N': 0, 'Si': 1, 'S': 1}
bin_cols = [
  'FAMI_TIENEINTERNET','FAMI_TIENELAVADORA',
  'FAMI_TIENEAUTOMOVIL','ESTU_PRIVADO_LIBERTAD',
  'ESTU_PAGOMATRICULAPROPIO','FAMI_TIENECOMPUTADOR'
]
for col in bin_cols:
    df[col + '_ENC'] = df[col].map(bin_map).astype(int)



In [ ]:
edu_order = [
  'Ninguno',
  'Primaria incompleta','Primaria completa',
  'Secundaria (Bachillerato) incompleta','Secundaria (Bachillerato) completa',
  'Técnica o tecnológica incompleta','Técnica o tecnológica completa',
  'Educación profesional incompleta','Educación profesional completa',
  'Postgrado',
  'No Aplica','No sabe','Desconocido'
]
ord_enc_padre = OrdinalEncoder(
    categories=[edu_order], dtype=int,
    handle_unknown='use_encoded_value', unknown_value=-1
)
df['EDU_PADRE_ENC'] = ord_enc_padre.fit_transform(
    df[['FAMI_EDUCACIONPADRE']]
)
df['EDU_MADRE_ENC'] = ord_enc_padre.fit_transform(
    df[['FAMI_EDUCACIONMADRE']]
)


In [ ]:
from sklearn.preprocessing import OrdinalEncoder

# 1) Definir y aplicar encoders ordinales
ord_encoders = {
  'VALOR_MATRICULA_ENC': ( ['ESTU_VALORMATRICULAUNIVERSIDAD'], matri_order ),
  'HORAS_TRABAJO_ENC':   ( ['ESTU_HORASSEMANATRABAJA'], horas_order ),
  'ESTRATO_ENC':         ( ['FAMI_ESTRATOVIVIENDA'], estrato_order ),
  'EDU_PADRE_ENC':       ( ['FAMI_EDUCACIONPADRE'], edu_order ),
  'EDU_MADRE_ENC':       ( ['FAMI_EDUCACIONMADRE'], edu_order )
}

for new_col, (cols, order) in ord_encoders.items():
    enc = OrdinalEncoder(
        categories=[order], dtype=int,
        handle_unknown='use_encoded_value', unknown_value=-1
    )
    df[new_col] = enc.fit_transform(df[cols])

# 2) Mapear binarios
for col in bin_cols:
    df[col + '_ENC'] = df[col].map(bin_map).astype(int)


In [ ]:
summary = pd.DataFrame({
    'column': df.columns,
    'unique_values': [df[col].unique().tolist() for col in df.columns]
})

# 3) Mostrarla en una sola vista
display(summary)

column  \
0                ESTU_VALORMATRICULAUNIVERSIDAD   
1                       ESTU_HORASSEMANATRABAJA   
2                          FAMI_ESTRATOVIVIENDA   
3                            FAMI_TIENEINTERNET   
4                           FAMI_EDUCACIONPADRE   
5                            FAMI_TIENELAVADORA   
6                           FAMI_TIENEAUTOMOVIL   
7                         ESTU_PRIVADO_LIBERTAD   
8                      ESTU_PAGOMATRICULAPROPIO   
9                          FAMI_TIENECOMPUTADOR   
10                          FAMI_EDUCACIONMADRE   
11                                       coef_1   
12                                       coef_2   
13                                       coef_3   
14                                       coef_4   
15                                     SEMESTRE   
16                                 PRGM_CLUSTER   
17                             PRGM_CLUSTER_SEM   
18                FACULTAD_Artes y Comunicación   
19    FACULTAD_Ciencias Naturales y Ambientales   
20       FACULTAD_Ciencias Sociales y Educación   
21  FACULTAD_Ciencias de la Salud y Veterinaria   
22               FACULTAD_Educación y Pedagogía   
23            FACULTAD_Ingeniería y Tecnologías   
24        FACULTAD_Negocios, Economía y Derecho   
25           FACULTAD_Otros/Interdisciplinarios   
26                              REGION_Amazonía   
27                                REGION_Andina   
28                                REGION_Caribe   
29                             REGION_Orinoquia   
30                              REGION_Pacífica   
31                                 REGION_otros   
32                          VALOR_MATRICULA_ENC   
33                            HORAS_TRABAJO_ENC   
34                                  ESTRATO_ENC   
35                       FAMI_TIENEINTERNET_ENC   
36                       FAMI_TIENELAVADORA_ENC   
37                      FAMI_TIENEAUTOMOVIL_ENC   
38                    ESTU_PRIVADO_LIBERTAD_ENC   
39                 ESTU_PAGOMATRICULAPROPIO_ENC   
40                     FAMI_TIENECOMPUTADOR_ENC   
41                                EDU_PADRE_ENC   
42                                EDU_MADRE_ENC   

                                        unique_values  
0   [Entre 5.5 millones y menos de 7 millones, Ent...  
1   [Menos de 10 horas, 0, Más de 30 horas, Entre ...  
2   [Estrato 3, Estrato 4, Estrato 5, Estrato 2, E...  
3                                            [Si, No]  
4   [Técnica o tecnológica incompleta, Técnica o t...  
5                                            [Si, No]  
6                                            [Si, No]  
7                                              [N, S]  
8                                            [No, Si]  
9                                            [Si, No]  
10  [Postgrado, Técnica o tecnológica incompleta, ...  
11  [0.322, 0.311, 0.297, 0.485, 0.316, 0.553, 0.2...  
12  [0.208, 0.215, 0.214, 0.172, 0.232, 0.142, 0.2...  
13  [0.31, 0.292, 0.305, 0.252, 0.285, 0.248, 0.26...  
14  [0.267, 0.264, 0.19, 0.294, 0.175, 0.308, 0.29...  
15                           [2021, 2020, 2019, 2018]  
16  [61, 6, 477, 379, 18, 538, 285, 436, 310, 405,...  
17  [36, 13, 57, 47, 65, 60, 46, 17, 73, 37, 27, 8...  
18                                         [0.0, 1.0]  
19                                         [0.0, 1.0]  
20                                         [0.0, 1.0]  
21                                         [1.0, 0.0]  
22                                         [0.0, 1.0]  
23                                         [0.0, 1.0]  
24                                         [0.0, 1.0]  
25                                         [0.0, 1.0]  
26                                         [0.0, 1.0]  
27                                         [0.0, 1.0]  
28                                         [0.0, 1.0]  
29                                         [0.0, 1.0]  
30                                         [0.0, 1.0]  
31  

In [ ]:
# Lista de columnas a remover (las que ya normalizaste y no usarás directamente)
cols_to_remove = [
    'ESTU_VALORMATRICULAUNIVERSIDAD',
    'ESTU_HORASSEMANATRABAJA',
    'FAMI_ESTRATOVIVIENDA',
    'FAMI_TIENEINTERNET',
    'FAMI_EDUCACIONPADRE',
    'FAMI_TIENELAVADORA',
    'FAMI_TIENEAUTOMOVIL',
    'ESTU_PRIVADO_LIBERTAD',
    'ESTU_PAGOMATRICULAPROPIO',
    'FAMI_TIENECOMPUTADOR',
    'FAMI_EDUCACIONMADRE'
]

# Elimina las columnas de tu DataFrame preprocesado (df)
df = df.drop(columns=cols_to_remove, errors='ignore')

# Verifica las columnas que quedan
print("Columnas restantes para el modelado:")
print(df.columns.tolist())


Columnas restantes para el modelado:
['coef_1', 'coef_2', 'coef_3', 'coef_4', 'SEMESTRE', 'PRGM_CLUSTER', 'PRGM_CLUSTER_SEM', 'FACULTAD_Artes y Comunicación', 'FACULTAD_Ciencias Naturales y Ambientales', 'FACULTAD_Ciencias Sociales y Educación', 'FACULTAD_Ciencias de la Salud y Veterinaria', 'FACULTAD_Educación y Pedagogía', 'FACULTAD_Ingeniería y Tecnologías', 'FACULTAD_Negocios, Economía y Derecho', 'FACULTAD_Otros/Interdisciplinarios', 'REGION_Amazonía', 'REGION_Andina', 'REGION_Caribe', 'REGION_Orinoquia', 'REGION_Pacífica', 'REGION_otros', 'VALOR_MATRICULA_ENC', 'HORAS_TRABAJO_ENC', 'ESTRATO_ENC', 'FAMI_TIENEINTERNET_ENC', 'FAMI_TIENELAVADORA_ENC', 'FAMI_TIENEAUTOMOVIL_ENC', 'ESTU_PRIVADO_LIBERTAD_ENC', 'ESTU_PAGOMATRICULAPROPIO_ENC', 'FAMI_TIENECOMPUTADOR_ENC', 'EDU_PADRE_ENC', 'EDU_MADRE_ENC']
